Source/Pawnmorphs/Esoteria/Genebank/ChamberDatabase.cs

```diff
@@ -0,0 +1,423 @@
+        /// </returns>
+        public bool CanAddToDatabase<T>([NotNull] GenebankEntry<T> entry)
+		{
+			return CanAddToDatabase(entry, out _);
+		}
+
+		/// <summary>
+		///     Determines whether this instance with the specified mutation definition can be added to the database
+		/// </summary>
+		/// <param name="entry">The genebank entry to check.</param>
+		/// <param name="reason">The reason.</param>
+		/// <returns>
+		///     <c>true</c> if this instance with the specified mutation definition  [can add to database]  otherwise, <c>false</c>
+		///     .
+		/// </returns>
+		public bool CanAddToDatabase<T>([NotNull] GenebankEntry<T> entry, out string reason)
+		{
+			if (entry == null) 
+                throw new ArgumentNullException(nameof(entry));
+
+            Type entryType = typeof(GenebankEntry<T>);
+			if (_genebankDatabase.ContainsKey(entryType) && _genebankDatabase[entryType].Contains(entry))
+			{
+				reason = ALREADY_TAGGED_REASON.Translate(entry.GetCaption());
+				return false;
+			}
+
+            int requiredStorage = entry.GetRequiredStorage();
+			if (FreeStorage < requiredStorage)
+			{
+				reason = NOT_ENOUGH_STORAGE_REASON.Translate(entry.GetCaption(), DatabaseUtilities.GetStorageString(requiredStorage), DatabaseUtilities.GetStorageString(FreeStorage));
+				return false;
+			}
+
+			if (!CanTag)
+			{
+				reason = NOT_ENOUGH_POWER.Translate();
+				return false;
+			}
+
+            if (entry.CanAddToDatabase(this, out reason) == false)
+            {
+                return false;
+            }
+
+			reason = "";
+			return true;
+		}
+
+
+
+		/// <summary>
+		///     Determines whether any of the specified mutation definitions can be
+		///     added to the database, and outputs an error if not.
+		/// </summary>
+		/// <param name="mutationDefs">The mutation definitions.</param>
+		/// <param name="reason">The reason the mutation cannot be ad.</param>
+		/// <returns>
+		///     <c>true</c> if at least one mutation definition can be added to database, otherwise <c>false</c>.
+		/// </returns>
+		public bool CanAddAnyToDatabase([NotNull] IEnumerable<MutationDef> mutationDefs, out string reason)
+		{
+			if (mutationDefs == null) throw new ArgumentNullException(nameof(mutationDefs));
+
+            MutationGenebankEntry entry = new MutationGenebankEntry(mutationDefs.First());
+            _genebankDatabase.TryGetValue(typeof(GenebankEntry<MutationDef>), out ExposedListContainer<IGenebankEntry> mutationEntries);
+            int? minRequiredCapacity = null;
+            MutationDef smallestMutation = null;
+			int neededCapacity = 0;
+            if (mutationEntries != null)
+            {
+                bool hasValid = false;
+                foreach (MutationDef mutationDef in mutationDefs)
+                {
+                    entry.Value = mutationDef;
+                    if (mutationEntries.Contains(entry) == false)
```

Comment:
You could avoid having to create the temporary entry here if you mapped mutationEntries to the values before the loop.  You could coalesce it to a set at the same time to make the multiple .Contains() calls faster, as well.

Something like this:
```
if (mutationEntries != null)
{
    bool hasValid = false;
    var mutationSet = mutationEntries.Select(x => x.Value).ToHashSet();
    foreach (MutationDef mutationDef in mutationDefs)
    {
        if (mutationSet.Contains(mutationdef) == false)
        {
            // ...
```
If you're comfortable using a more functional style, you could also redo this whole block with it:

```
var smallestMutation = mutationEntries.Select(x => x.Value)
        .Union(mutationDefs)
        .MinBy(x => x.GetRequiredStorage());

if (smallestMutation == null)
{
    // All already tagged.
    reason = ALREADY_TAGGED_MULTI_REASON.Translate();
    return true;
}
```

Reply:
Right because it is specifically mutations which have requiredStorage as extension method so we don't need to use abstraction.

Optimally that function would be located on the abstracted object with any additional requirements to CanAdd.